File: SegmentTree/MaximumSumQuery.java
import java.util.*;
import java.io.*;

class Node{
    long maxSum,totalSum,maxSuffixSum,maxPrefixSum;
    public Node(){maxSum=0;totalSum=0;maxSuffixSum=0;maxPrefixSum=0;}
    public void setLeaf(int value){maxSum=value;totalSum=value;maxSuffixSum=value;maxPrefixSum=value;}
    public void setCompare(Node node1,Node node2){
        maxSum = Math.max(Math.max(Math.max(node1.maxSum,node2.maxSum),
                                  Math.max(node1.totalSum+node2.maxPrefixSum,node1.maxSuffixSum+node2.totalSum)
                                  ),node1.maxSuffixSum+node2.maxPrefixSum);
        totalSum = node1.totalSum + node2.totalSum;
        maxPrefixSum = Math.max(node1.maxPrefixSum,node1.totalSum+node2.maxPrefixSum);
        maxSuffixSum = Math.max(node2.maxSuffixSum,node2.totalSum+node1.maxSuffixSum);
    }
}

class SegmentTree{
    int size;
    Node tree[] = null;
    
    public SegmentTree(int[] array){
        size = array.length;
        tree = new Node[4*size];
        for(int i=0;i<tree.length;i++)
            tree[i] =new Node();
        buildTree(0,size-1,1,array);
    }
    
    private void buildTree(int left,int right,int treeNode,int[] array){
        if(left==right){
            tree[treeNode].setLeaf(array[left]);
            return;
        }
        int mid = (left+right)/2;
        buildTree(left,mid,2*treeNode,array);
        buildTree(mid+1,right,2*treeNode+1,array);
        tree[treeNode].setCompare(tree[2*treeNode],tree[2*treeNode+1]);
    }
    
    public long query(int minIndex,int maxIndex){
        Node ansNode = queryTree(0,size-1,1,minIndex,maxIndex);
        return ansNode.maxSum;
    }
    
    private Node queryTree(int left,int right,int treeIndex,int arrayMinIndex,int arrayMaxIndex){
        if(arrayMinIndex>right || arrayMaxIndex<left)
            return null;
        else if(arrayMinIndex<=left && right<=arrayMaxIndex)
            return tree[treeIndex];
        int mid = (left+right)/2;
        Node nodeLeft = queryTree(left,mid,2*treeIndex,arrayMinIndex,arrayMaxIndex);
        Node nodeRight = queryTree(mid+1,right,2*treeIndex+1,arrayMinIndex,arrayMaxIndex);
        return setCompareReturnNode(nodeLeft,nodeRight);
    }
    
    private Node setCompareReturnNode(Node node1,Node node2){
        if(node1==null && node2==null)
            return null;
        else if(node1==null)
            return node2;
        else if(node2==null)
            return node1;
        Node node = new Node();
        node.setCompare(node1,node2);
        return node;
    }
    
    public void printTree(){
        System.out.println("---- Tree ----");
        for(int i=0;i<tree.length;i++)
            System.out.println(tree[i].maxSum+" "+tree[i].totalSum+" "+tree[i].maxPrefixSum+" "+tree[i].maxSuffixSum);
        System.out.println("---- ---- ----");
    }
}
public class Main {
	public static void main(String[] args) {
		// Write your code here
        FastReader reader = new FastReader();
        int elements = reader.nextInt();
        int array[] = new int[elements];
        for(int i=0;i<elements;i++)
            array[i] = reader.nextInt();
        int queries = reader.nextInt();
        SegmentTree segTree = new SegmentTree(array);
        // segTree.printTree();
        while(queries-->0){
            int minIndex = reader.nextInt();
            int maxIndex = reader.nextInt();
            System.out.println(segTree.query(minIndex-1,maxIndex-1));
        }
	}

    static class FastReader { 
        BufferedReader br; 
        StringTokenizer st; 
  
        public FastReader() { 
            br = new BufferedReader(newInputStreamReader(System.in)); 
        } 
  
        String next() { 
            while (st == null || !st.hasMoreElements()) { 
                try{ 
                    st = new StringTokenizer(br.readLine()); 
                } catch (IOException  e) { 
                    e.printStackTrace(); 
                } 
            } 
            return st.nextToken(); 
        } 
  
        int nextInt() { 
            return Integer.parseInt(next()); 
        } 
  
        long nextLong() { 
            return Long.parseLong(next()); 
        } 
  
        double nextDouble() { 
            return Double.parseDouble(next()); 
        } 
  
        String nextLine() { 
            String str = ""; 
            try{ 
                str = br.readLine(); 
            } catch (IOException e) { 
                e.printStackTrace(); 
            } 
            return str; 
        } 
    }
}

File: SegmentTree/MaximumSumQuery.cpp
#include <iostream>
#include <cmath>
#include <vector>
#define ll long long int
using namespace std;

class Node{
public:
    ll maxValue,maxFromLeft,maxFromRight,totalSum;

    Node(){
        setValue(0);
    }

    void setValue(ll value){
        maxValue = value;totalSum=value;maxFromLeft=value;maxFromRight=value;totalSum=value;
    }

    void setNodeValues(Node* nodeLeft,Node* nodeRight){
        if(nodeLeft!=NULL && nodeRight!=NULL){
            totalSum = nodeLeft->totalSum+nodeRight->totalSum;
            maxFromLeft = max(nodeLeft->maxFromLeft,nodeLeft->totalSum+nodeRight->maxFromLeft);
            maxFromRight = max(nodeRight->maxFromRight,nodeRight->totalSum+nodeLeft->maxFromRight);
            maxValue = max(nodeLeft->totalSum,nodeRight->totalSum,totalSum,maxFromLeft,maxFromRight,nodeLeft->maxFromRight+nodeRight->maxFromLeft,nodeLeft->maxValue,nodeRight->maxValue);
        } else if(nodeLeft.initialized){}
    }

    ll max(ll a,ll b){
        if(a>=b)
            return a;
        return b;
    }

    ll max(ll a,ll b,ll c,ll d,ll e,ll f,ll g,ll h){
        return max(max(max(max(a,b),max(c,d)),max(e,f)),max(g,h));
    }
};

class SegmentTree{
private:
    vector<Node> tree;
    ll elements;

    void buildTree(ll low,ll high,ll treeIndex,ll* array){
        if(low==high){
            tree[treeIndex].setValue(array[low]);
            return;
        }
        ll mid = (low+high)/2;
        buildTree(low,mid,2*treeIndex+1,array);
        buildTree(mid+1,high,2*treeIndex+2,array);
        tree[treeIndex].setNodeValues(&tree[2*treeIndex+1],&tree[2*treeIndex+2]);
    }

    Node queryTree(ll low,ll high,ll treeLowIndex,ll treeHighIndex,ll treeIndex){
        Node resultNode;
        if(treeLowIndex>high || treeHighIndex<low)
            return resultNode;
        else if(low<=treeLowIndex && treeHighIndex<=high)
            return tree[treeIndex];
        ll mid = (treeLowIndex+treeHighIndex)/2;
        Node leftResultSet = queryTree(low,high,treeLowIndex,mid,2*treeIndex+1);
        Node rightResultSet = queryTree(low,high,mid+1,treeHighIndex,2*treeIndex+2);
        resultNode.setNodeValues(&leftResultSet,&rightResultSet);
        return resultNode;
    }

public:
    SegmentTree(ll* array,ll elements){
        this->elements = elements;
        tree = new Node[4*elements+3];
        buildTree(0,elements-1,0,array);
    }

    ll query(ll leftIndex,ll rightIndex){
        Node resultNode = queryTree(leftIndex,rightIndex,0,elements-1,0);
        return resultNode.maxValue;
    }

    void printTree(){
        cout<<"maxValue: ";
        for(ll i=0;i<4*elements+3;i++)
            cout<<tree[i].maxValue<<" "<<tree[i].totalSum<<" "<<tree[i].maxFromLeft<<" "<<tree[i].maxFromRight<<endl;
    }
};

int main(){
    ll elements;
    cin>>elements;
    ll* array = new ll[elements];
    for(ll i=0;i<elements;i++)
        cin>>array[i];
    SegmentTree segmentTree(array,elements);
    // segmentTree.printTree();
    ll queries;
    cin>>queries;
    while(queries-->0){
        ll leftIndex,rightIndex;
        cin>>leftIndex>>rightIndex;
        cout<<segmentTree.query(leftIndex-1,rightIndex-1)<<endl;
        // segmentTree.printTree();
    }
}